### ISSUE TITLE
Migrator rerunning migrations for testing when nothing added

### ISSUE THREAD

**Issue by mcube27** [cakephp/migrations#533]
Cakephp 4.3.6
Migrations 3.4.0

When I run my test, I do :

```
$migrator = new Migrator();
$migrator->run();
```
Not adding anything to my migrations, everytime I run tests, the cli-debug log writes :
```
2022-03-09 08:15:14 debug: Reading migrations status for test...
2022-03-09 08:15:14 debug: One or more additional migrations detected.
```
And rerun the migrations, causing missing table and other problems.

We have a quite big application with a log of migrations and I can't seem to see another explaination for this.

**Comment by mcube27** [comment]
Proposed fix :

in "vendor/cakephp/migrations/src/TestSuite/Migrator.php:165"

Change to `return false;` 

Since I did this, I have no problems

**Comment by markstory** [comment]
I don't think your proposed fix is a proper solution. Changing that return value will result in migration state not being reset when there are new migrations that need to be applied.

We could improve the debug output to include the name of the migration that status checks believe to be incorrect.

> And rerun the migrations, causing missing table and other problems.

Shouldn't migrations create all the migrations for your application?

**Comment by mcube27** [comment]
Hi,

Thank you for your fast answer !

My understanding is :
- My test database is populated with the migrations. As is my production database.
- It uses the same phynxlog table to know which migrations is up, down or missing.

The expected behavior for me is that it uses this information to migrate the data.

What I have seen so far :
- it passes through a function that decides what needs to be migrated
- it migrates my database at each beginning of my tests (not since I corrected as above)
- it generates errors as some tables as not existing anymore

Solutions :
- better debug would be great as there is little knowledge with the actual debug to know where somethings wrong
- let the phynxlog be used
- allow parameters as "date" or "target" to be used in the Migrator::run, I tested it but doesn't seem to work. It allows debugging or working on a database before a migration.

Thank you very much for your help and at your disposal for any complementary information!

**Comment by markstory** [comment]
> it generates errors as some tables as not existing anymore

Where do the tables go though? Do you have tests that drop tables?

**Comment by mcube27** [comment]
What I meant is that if Migrations are rerun, some tables created in the begining of our app are no longer existing in the actual state of the app.

Not easy to explain.

In my tests I do not drop tables.

But I get very frustating errors and they are not always identical as if during migrations something goes wrong but the migration passes anyway.

I tried going for the sql dump, but not good either.

**Comment by liqueurdetoile** [comment]
Hi !

Following @pabloelcolombiano advice, I've switched today to the embedded migrator for tests (previously using vierge-noire one).

I'm posting here because I think my trouble is very similar as migrator keeps remigrating each time I'm launching tests. I'm using `cakephp/migrations~3.5.1`.

I think the trouble is here : https://github.com/cakephp/migrations/blob/9b12dcd44a53b7607386ffa3eb2b7845e6a84d48/src/TestSuite/Migrator.php#L168

I can't see why an already applied migration should be considered as unapplied ?

When commenting the line, behavior resumed to correct. Migration is only done if not already up.

**Comment by pabloelcolombiano** [comment]
@liqueurdetoile thanks for reporting this here. I am preparing a PR for that.

**Comment by Rakasch** [comment]
Same problem here.

I have just updated to Cake4.3.
Since I have not used any migrations in the project before, I now only have an initial migration with the already existing DB.
Each time TestSuite is started, all the tables are all truncated and recreated.
I debugged through the code to see where and why this is happening.

I found that shouldDropTables for my migration is always "up".
src/TestSuite/Migrator.php
![image](https://user-images.githubusercontent.com/8471027/167398298-63e9971e-8d30-4b3d-a1fc-4df12dd95c36.png)

That's because it gets its status from here:
src/CakeManager.php
![image](https://user-images.githubusercontent.com/8471027/167398415-73eba6b7-0a44-4354-8752-e6e9b2b38f02.png)

Currenlty it says "up" if it's already migrated. But if the migration already is in the version log, nothing needs to be migrated anymore. The shouldDropTables() function should not migrate it again, when it's already "up".
Maybe the problem is, that printStatus returns "up" in two different cases. When it's missing AND when it's already up.
